scheduler/pkg/agent/client.go

```diff
@@ -230,66 +264,86 @@ func (c *Client) createConnection() error {
 	return nil
 }
 
-func (c *Client) WaitReady() error {
+func (c *Client) WaitReadySubServices(isStartup bool) error {
 	logger := c.logger.WithField("func", "waitReady")
 
+	backoffWithMax := backoff.NewExponentialBackOff()
+	if isStartup {
+		backoffWithMax.MaxElapsedTime = c.settings.maxElapsedTimeReadySubServiceBeforeStart
+	} else {
+		backoffWithMax.MaxElapsedTime = c.settings.maxElapsedTimeReadySubServiceAfterStart
+	}
+
 	// Wait for model repo to be ready
 	logFailure := func(err error, delay time.Duration) {
 		logger.WithError(err).Errorf("Rclone not ready")
 	}
-	logger.Infof("Waiting for Model Repository to be ready")
+
 	//TODO make retry configurable
-	err := backoff.RetryNotify(c.ModelRepository.Ready, backoff.NewExponentialBackOff(), logFailure)
+	err := backoff.RetryNotify(c.ModelRepository.Ready, backoffWithMax, logFailure)
 	if err != nil {
 		logger.WithError(err).Error("Failed to wait for model repository to be ready")
 		return err
 	}
 
-	// Wait for V2 Server to be ready
+	// Wait for Inference server to be ready
 	logFailure = func(err error, delay time.Duration) {
-		logger.WithError(err).Errorf("Server not ready")
+		logger.WithError(err).Errorf("Inference server not ready")
 	}
-	logger.Infof("Waiting for inference server to be ready")
-	err = backoff.RetryNotify(c.stateManager.v2Client.Ready, backoff.NewExponentialBackOff(), logFailure)
+
+	err = backoff.RetryNotify(c.stateManager.v2Client.Live, backoffWithMax, logFailure)
 	if err != nil {
 		logger.WithError(err).Error("Failed to wait for inference server to be ready")
 		return err
 	}
 
-	// Unload any existing models on server to ensure we start in a clean state
-	logger.Infof("Unloading any existing models")
-	err = c.UnloadAllModels()
-	if err != nil {
-		return err
+	if isStartup {
+		// Unload any existing models on server to ensure we start in a clean state
+		logger.Infof("Unloading any existing models")
+		err = c.UnloadAllModels()
+		if err != nil {
+			return err
+		}
 	}
 
 	// http reverse proxy
-	if err := startSubService(c.rpHTTP, logger); err != nil {
-		logger.WithError(err).Error("Failed to start http proxy")
+	if err := isReadyChecker(
+		isStartup, c.rpHTTP, logger, "Rest proxy not ready",
+		c.settings.maxElapsedTimeReadySubServiceBeforeStart,
+		c.settings.maxElapsedTimeReadySubServiceAfterStart,
+	); err != nil {
 		return err
 	}
 
 	// grpc reverse proxy
-	if err := startSubService(c.rpGRPC, logger); err != nil {
-		logger.WithError(err).Error("Failed to start grpc proxy")
+	if err := isReadyChecker(isStartup, c.rpGRPC, logger, "Grpc proxy not ready",
```

Comment:
Not sure I follow why internall services need to be rechecked for readiness?

Reply:
As these internal services are really running as separate go routines, I am assuming that there could be errors that cause them to stop for any reason? TBH I have not spotted any error of this kind yet though.

Reply:
Could the code be simplified by setting a single var for time rather than passing 2 parameters to each call?

Reply:
done in [af21229](https://github.com/SeldonIO/seldon-core/pull/4611/commits/af2122998018e958d2d970fd8f0b6913344ae2c7)